### PR TITLE
Introduce TestLogger

### PR DESCRIPTION
**Pull request by antocuni**
In PR #58481 we discussed about introducing a test logger, to allow tests to log important information which are displayed in case of failure. I think it's easier to discuss the details in a fresh and separate PR, so here is a first design attempt.

The `TestLogger` interface is super-simple for now: each test gets a fresh instance of `self.logger`, and test writers just need to use `self.logger.print()`. We can decide to add more specific methods in the future, e.g. to address #60748.

I decided to use `logger.print()` instead of e.g. `logger.log()` to underline the fact that it has the same signature of the builtin `print()`. It happens too often in other pieces of software to find a `.log()` method and you never know what is the exact signature.

Example of output:
```python
    def test_logger_example(self):
        self.logger.print('hello world')
        for i in range(5):
            self.logger.print('i =', i)
            x = 10 / (3-i)
```

```
 $ python test/test_testing.py -k logger_example
EE
======================================================================
ERROR: test_logger_example_cpu (__main__.TestTestingCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/antocuni/git/Quansight/pytorch/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "test/test_testing.py", line 629, in test_logger_example
    x = 10 / (3-i)
ZeroDivisionError: division by zero
~~~~~~~~ TestLogger ~~~~~~~~
    hello world
    i = 0
    i = 1
    i = 2
    i = 3
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

======================================================================
ERROR: test_logger_example_meta (__main__.TestTestingMETA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/antocuni/git/Quansight/pytorch/torch/testing/_internal/common_utils.py", line 1046, in wrapper
    method(*args, **kwargs)
  File "/home/antocuni/git/Quansight/pytorch/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "test/test_testing.py", line 629, in test_logger_example
    x = 10 / (3-i)
ZeroDivisionError: division by zero
~~~~~~~~ TestLogger ~~~~~~~~
    hello world
    i = 0
    i = 1
    i = 2
    i = 3
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

----------------------------------------------------------------------
Ran 2 tests in 0.002s

FAILED (errors=2)
```

It is important to note that the content it's attached to the original exception message, thus making sure that it's displayed in the right position. At first I tried to simply `print()` the logger's content to `stderr` in case of exception, but with `unittest` it's not easy to map the `stderr` to each failing test. The output looked like this, so the approach taken by this PR is clearly superior:

```
$ python test/test_testing.py -k logger_example
~~~~~~~~ TestLogger ~~~~~~~~
    hello world
    i = 0
    i = 1
    i = 2
    i = 3
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

E~~~~~~~~ TestLogger ~~~~~~~~
    hello world
    i = 0
    i = 1
    i = 2
    i = 3
~~~~~~~~~~~~~~~~~~~~~~~~~~~~


E
======================================================================
ERROR: test_logger_example_cpu (__main__.TestTestingCPU)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/antocuni/git/Quansight/pytorch/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "test/test_testing.py", line 629, in test_logger_example
    x = 10 / (3-i)
ZeroDivisionError: division by zero

======================================================================
ERROR: test_logger_meta (__main__.TestTestingMETA)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/antocuni/git/Quansight/pytorch/torch/testing/_internal/common_device_type.py", line 376, in instantiated_test
    result = test(self, **param_kwargs)
  File "test/test_testing.py", line 629, in test_logger_example
    x = 10 / (3-i)
ZeroDivisionError: division by zero

----------------------------------------------------------------------
Ran 2 tests in 0.002s

FAILED (errors=2)
```

cc @mruberry

### CODE REVIEW

File: torch/testing/_internal/common_device_type.py [hunk 1]
```diff
@@ -368,17 +369,21 @@ def instantiated_test(self, param_kwargs=param_kwargs):
                 # Note: precision is reset after the test is run
                 guard_precision = self.precision
                 guard_rel_tol = self.rel_tol
+                self.logger = TestLogger()
```
Comment (mruberry) [inline torch/testing/_internal/common_device_type.py:372]: This is a cool approach

cc @jbschlosser who's been looking at our test architecture recently

**Comment by mruberry** [comment]
This seems very smart and cool to me, @antocuni!

@pmeier and @jbschlosser, would you like to take a look, too?

**Review by pmeier — COMMENTED** [review]
@antocuni 

> allow tests to log important information which are displayed in case of failure.

Could you be more specific? What kind of failure are we talking about? In case comparison errors, for example `AssertionError`'s raised by `TestCase.assertEqual`, I would prefer to simply use a custom error message.

**Comment by antocuni** [comment]
> @antocuni
> Could you be more specific? What kind of failure are we talking about? In case comparison errors, for example `AssertionError`'s raised by `TestCase.assertEqual`, I would prefer to simply use a custom error message.

the idea is that each test can decide what to log, and that logged infos are useful to inspect in case of failure. But this is up to the test to decide, and I don't think that `TestCase.assertEqual` should have the possibility to hide them.

The first concrete use case that we have is #58481 : the idea is that each test can do `self.logger.print(sample.name)`, so that it will be obvious which sample is causing the failure. If `TestCase.assertEqual` hides the log, then we are back to square 1.

**Comment by pmeier** [comment]
> the idea is that each test can do self.logger.print(sample.name), so that it will be obvious which sample is causing the failure.

But that is exactly what a custom error message is for:

```python
self.assertEqual(..., msg=f"Sample {sample.name} caused the failure.")
```

So far I fail to see the added benefit of having the logger. Happy to be convinced otherwise.

**Comment by antocuni** [comment]
> But that is exactly what a custom error message is for:
> 
> ```python
> self.assertEqual(..., msg=f"Sample {sample.name} caused the failure.")
> ```

only if you remind to to write the custom message on every assert of every test. The point is precisely NOT to have to write it explicitly, because the status quo proves that people don't do it.
Moreover, there might be other exceptions which don't have a custom message.

**Comment by pmeier** [comment]
> only if you remind to to write the custom message on every assert of every test. The point is precisely NOT to have to write it explicitly, because the status quo proves that people don't do it.

From your example it seems that the `self.logger.print()` call also has to be written manually. Could you provide an example where using a custom error message is either harder or conveys less information than the logger approach?

> Moreover, there might be other exceptions which don't have a custom message.

The only way I see `assertEqual` ignoring a custom error message is if it fails due to some unexpected inputs. That shouldn't happen unless you currently write the test or work on `assertEqual`. Still, I don't see a special sample fail in such a case, but all of them. Do you have an example to better judge this?

**Comment by jbschlosser** [comment]
I'd use this for interactively debugging test failures, since it's clearly superior to naive print-based debugging. For this use case, this approach is more flexible than assert messages since anything can be logged at any time. I'd probably remove calls to the logger before actually checking in the test, however.

**Comment by mruberry** [comment]
The two reasons I like this are:

- crafting super involved error messages for every assertion is, empirically, a pain, or we would see more elaborate error messages much more often
- tests can fail outside an assertion and then that elaborate error message never shows up, anyway

So I think this is an easier way to write helpful debug information that's more reliable than the assertions.

**Comment by pmeier** [comment]
@jbschlosser 

> I'd use this for interactively debugging test failures, since it's clearly superior to naive print-based debugging. For this use case, this approach is more flexible than assert messages since anything can be logged at any time. I'd probably remove calls to the logger before actually checking in the test, however.

I can get behind that use case. As @antocuni already kinda mentioned above, `pytest` has this feature built-in:

```python
def test_logger_example():
    print('hello world')
    for i in range(5):
        print('i =', i)
        x = 10 / (3 - i)
```

```
__________________________________ test_logger_example __________________________________
Traceback (most recent call last):
  File "/home/philip/git/pytorch/torch/test_logger.py", line 5, in test_logger_example
    x = 10 / (3 - i)
ZeroDivisionError: division by zero
--------------------------------- Captured stdout call ----------------------------------
hello world
i = 0
i = 1
i = 2
i = 3

```

Thus, the logger is only needed for working with `unittest`. 

@mruberry 

> crafting super involved error messages for every assertion is, empirically, a pain, or we would see more elaborate error messages much more often

Maybe I'm missing something, but how is `self.logger.print("i =", i)` any easier than `self.assertEqual(..., msg=f"i={i}")`? If very simple messages are ok for the logger why do we require "super involved" or "elaborate" ones for the assertions? 

I feel like people don't write custom error messages, because it is more effort that not immediately pays off. But since writing logger calls is a similar effort, I don't see that changing in the future.

> tests can fail outside an assertion and then that elaborate error message never shows up, anyway

In that case we are back at the interactive debugging, which I can get behind.

---

To conclude, I'm not opposed to this feature, but I think it only has a very narrow use case. It is useful for interactive debugging with `unittest` as runner, but as a `pytest` user I don't need it. 

If we go for it, I agree with @jbschlosser to treat the logger calls the same as debug print statements and remove them before checking in the code.

**Comment by mruberry** [comment]
No but if I pass the the msg argument like that to assertEqual then don't I clobber the default message?

Also, yes, in the example you're showing the msg is simple, but if one wants to build a complicated error message incrementally and preserve the default assertion message then that requires writing to a buffer and then adding some functionality to assertEqual to preserve the default msg and preserve the buffer.

**Comment by pmeier** [comment]
> No but if I pass the the msg argument like that to assertEqual then don't I clobber the default message?

Yes, you would. That is why I proposed 5. in #58383. Maybe we should reconsider it? Still, that would make the call more elaborate. 

> Also, yes, in the example you're showing the msg is simple,

Well, that was the example given to motivate the need for the logger. Happy to discuss this on a more elaborate example.

> but if one wants to build a complicated error message incrementally

Ideally, each test should test exactly one thing and thus feature one assertion. In cases where this is not possible, the assertions should be either independent of each other or at least organized in a way that makes earlier error messages obsolete in case a later check fails:

```python
if input.grad is None:
    raise AssertionError("Input has no grad")

self.assertEqual(input.grad, expected_grad, msg="Gradients mismatch")
```

Can you provide an example where we this is not possible and thus an incrementally build error message would be helpful?

**Comment by antocuni** [comment]
> Well, that was the example given to motivate the need for the logger. Happy to discuss this on a more elaborate example.

@pmeier consider this real world test failure (which was the one of the reasons which motivated me to start this): https://github.com/pytorch/pytorch/pull/53119#issuecomment-810548792

We had a test which failed only on ROCm, so no easy way to reproduce it locally. And it failed deep inside `gradcheck`: as you see, no custom error message can help in that case, and `gradcheck` doesn't even have the notion of what is a sample input.

#58481 is an attempt to fix the problem, make our test suite better and make it easier to debug these kind of problems in the future. That PR tried various ways to handle it and it was decided that a TestLogger was the best way to move forward because it can be used also for other things (e.g. #60748). That's how we ended up here.

I agree that it's functionally equivalent to using `print()+pytest`, but our test suite uses `unittest` so we have to live with that :man_shrugging:

**Comment by pmeier** [comment]
> as you see, no custom error message can help in that case, and gradcheck doesn't even have the notion of what is a sample input.

Why not? Instead of using a logger, you could just simply put a custom error message [here](https://github.com/antocuni/pytorch/blob/94d639b31eddc98fa8faca76f508191dcb5e7ce2/test/test_ops.py#L124-L127).

But anyway, I see there is a use case for interactive debugging and without some improvements for the custom error message outlined in #58383 (5), a custom error message might be harder to do. Since it seems that in #58481 it was already decided to go for this, let's move forward.

**Comment by mruberry** [comment]
Just checking in @antocuni. This LGTM! Is there anything we should do before it's merged? If it's ready to go would you just rebase this and ping me?

**Comment by mruberry** [comment]
@jbschlosser would you like to add a comment before this is merged?

**Review by mruberry — APPROVED** [review]

**Comment by antocuni** [comment]
> Just checking in @antocuni. This LGTM! Is there anything we should do before it's merged? If it's ready to go would you just rebase this and ping me?

done! A few tests are still running at the moment of writing but most of them are passing, so it should be good to merge

**Review by jbschlosser — APPROVED** [review]
I'm good with this- thanks for adding it!

**Comment by mruberry** [comment]
FYI we'll have to resolve an internal issue with test_testing.TestTestingCPU which is failing with

```
AssertionError: 'Exception: failing on purpose' not found in 'Traceback (most recent call last):\n  File "<string>", line 4, in <module>\nModuleNotFoundError: No module named \'torch.testing\'\n'
```

This might mean we just need to update the internal test dependencies to depend on torch.testing, although other tests in test_testing.py rely on the module, too, so I'm not sure how they're working.

**Comment by antocuni** [comment]
> FYI we'll have to resolve an internal issue with test_testing.TestTestingCPU which is failing with
> 
> ```
> AssertionError: 'Exception: failing on purpose' not found in 'Traceback (most recent call last):\n  File "<string>", line 4, in <module>\nModuleNotFoundError: No module named \'torch.testing\'\n'
> ```
> 
> This might mean we just need to update the internal test dependencies to depend on torch.testing, although other tests in test_testing.py rely on the module, too, so I'm not sure how they're working.

Just to make sure to understand correctly: this is something which which needs to be fixed by the facebook team and there is no action needed from my side, right?

**Comment by mruberry** [comment]
> Just to make sure to understand correctly: this is something which which needs to be fixed by the facebook team and there is no action needed from my side, right?

Correct; sorry that was unclear.

**Pull request closed without merging**